pkg/ottl/ottlfuncs/func_parse_csv.go

```diff
@@ -0,0 +1,229 @@
+	}
+
+	return nil, fmt.Errorf("unknown mode: %s", mode)
+}
+
+func parseCSV[K any](target, header ottl.StringGetter[K], delimiter, headerDelimiter rune, lazyQuotes bool) ottl.ExprFunc[K] {
+	headerDelimiterString := string([]rune{headerDelimiter})
+
+	return func(ctx context.Context, tCtx K) (any, error) {
+		targetStr, err := target.Get(ctx, tCtx)
+		if err != nil {
+			return nil, fmt.Errorf("get target: %w", err)
+		}
+
+		headerStr, err := header.Get(ctx, tCtx)
+		if err != nil {
+			return nil, fmt.Errorf("get header: %w", err)
+		}
+
+		if headerStr == "" {
+			return nil, errors.New("headers must not be an empty string")
+		}
+
+		headers := strings.Split(headerStr, headerDelimiterString)
+
+		csvReader := csv.NewReader(strings.NewReader(targetStr))
+		csvReader.Comma = delimiter
+		csvReader.FieldsPerRecord = len(headers)
+		csvReader.LazyQuotes = lazyQuotes
+
+		fields, err := csvReadLine(csvReader)
+		if err != nil {
+			return nil, err
+		}
+
+		return csvHeadersMap(headers, fields)
+	}
+}
+
+func parseCSVIgnoreQuotes[K any](target, header ottl.StringGetter[K], delimiter, headerDelimiter rune) ottl.ExprFunc[K] {
+	headerDelimiterString := string([]rune{headerDelimiter})
+	delimiterString := string([]rune{delimiter})
+
+	return func(ctx context.Context, tCtx K) (any, error) {
+		targetStr, err := target.Get(ctx, tCtx)
+		if err != nil {
+			return nil, fmt.Errorf("get target: %w", err)
+		}
+
+		headerStr, err := header.Get(ctx, tCtx)
+		if err != nil {
+			return nil, fmt.Errorf("get header: %w", err)
+		}
+
+		if headerStr == "" {
+			return nil, errors.New("headers must not be an empty string")
+		}
+
+		headers := strings.Split(headerStr, headerDelimiterString)
+
+		// Ignoring quotes makes CSV parseable with just string.Split
+		fields := strings.Split(targetStr, delimiterString)
+		return csvHeadersMap(headers, fields)
+	}
+}
+
+// csvHeadersMap creates a map of headers[i] -> fields[i].
+func csvHeadersMap(headers []string, fields []string) (pcommon.Map, error) {
+	pMap := pcommon.NewMap()
+	parsedValues := make(map[string]any)
+
+	if len(fields) != len(headers) {
+		return pMap, fmt.Errorf("wrong number of fields: expected %d, found %d", len(headers), len(fields))
+	}
+
+	for i, val := range fields {
+		parsedValues[headers[i]] = val
+	}
+
+	err := pMap.FromRaw(parsedValues)
+	if err != nil {
+		return pMap, fmt.Errorf("create pcommon.Map: %w", err)
+	}
+
+	return pMap, nil
+}
+
+// csvReadLine reads a CSV line from the csv reader, returning the fields parsed from the line.
+// We make the assumption that the payload we are reading is a single row, so we allow newline characters in fields.
+// However, the csv package does not support newlines in a CSV field (it assumes rows are newline separated),
+// so in order to support parsing newlines in a field, we need to stitch together the results of multiple Read calls.
+func csvReadLine(csvReader *csv.Reader) ([]string, error) {
```

Comment:
In a similar PR (https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31035#discussion_r1481919680) we discussed sharing parsing functionality between pkg/stanza and OTTL. Would the same make sense here?

Reply:
Reusing as much as we can from https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/stanza/operator/parser/csv/csv.go feels appropriate.